### PR TITLE
Enable stricter tslint rules

### PR DESCRIPTION
**Pull request by jLouzado**
## Changes

Closes: #18 

## Dependancies:

#28 - Rename files to camelCase

## Remaining TODOs
- [x] `only-arrow-functions` enable 
- [ ] match.ts
- [ ] commandFunction.ts
- [x] reducerFunction.ts

### CODE REVIEW

File: modules/core/src/action.ts [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
- * @param {any} value
- * @returns {Action}
+ * @param: Action Type
+ * @param: Action Payload
  */
 export const action: {
-  <T>(type: string | number, value: T): Action<T>
-  <T>(type: string | number): {(value: T): Action<T>}
+  <T>(type: string | number, value: T): IAction<T>
+  <T>(type: string | number): (value: T) => IAction<T>
```
Comment (jLouzado) [inline modules/core/src/action.ts]: even though `createAction` has good typings now, the inferred return type is overwritten here.
```typescript
const actual = action('WIND', Nil()) // IAction<INilAction>
```

Ref: https://github.com/action-land/action-land/pull/29/files#diff-ade9c5c4a4ecd2bb9c3785ac934117e2R26

File: modules/smitten/src/defaultEmitter.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  public constructor(
+    public readonly type: string | number,
+    public readonly parent: DefaultEmitter | RootEmitter
+  ) {}
+  public emit = (value: EmitValue): void => {
+    if (this instanceof RootEmitter) {
+      this.emit(value)
+    } else {
+      let current = this.parent
+      let act = action(this.type, value)
```
Comment (jLouzado) [inline modules/smitten/src/defaultEmitter.ts]: - Tests pass even if `act` is initialized to `value`
  - That seems like it'd be a bug though
- I think a test-case is missing but I'm not able to write a failing test. 🙈

File: tslint.json [hunk 3]
```diff
@@ -1,40 +1,30 @@
...
-      "check-type"
+    "variable-name": [true, "allow-pascal-case"],
+    "typedef": [true, "call-signature"],
+    "file-name-casing": [true, "camel-case"],
+    "no-implicit-dependencies": [
+      false, // Doesn't play well with mono-repo
+      "dev"
     ],
-    "ordered-imports": true
+    "only-arrow-functions": false,
```
Comment (fenilkanjani) [inline tslint.json]: This should not be entirely false

File: tslint.json [hunk 4]
```diff
@@ -1,40 +1,30 @@
...
+    "increment-decrement": false,
+    "no-null-keyword": false,
+    "no-magic-numbers": false,
+    "no-parameter-properties": false,
+    "no-unused-expression": true
```
Comment (fenilkanjani) [inline tslint.json]: Why are last 6 set to be false?

File: test/smitten/smitten.ts [hunk 5]
```diff
@@ -48,7 +50,7 @@ describe('hoe', () => {
     const f = e.of('F')
     e.emit.call(null, 100)
     f.emit.call(null, 200)
-    assert.deepEqual(actions, [100, action('F', 200)])
+    assert.deepStrictEqual(actions, [100, action('F', 200)])
```
Comment (fenilkanjani) [inline test/smitten/smitten.ts]: how is this part of lint changes?

File: tsconfig.json [hunk 6]
```diff
@@ -2,6 +2,17 @@
   "compilerOptions": {
     "target": "es6",
     "module": "commonjs",
-    "strict": true
+    "strict": true,
+    "plugins": [
+      {
+        "name": "typescript-tslint-plugin",
+        "alwaysShowRuleFailuresAsWarnings": false,
+        "ignoreDefinitionFiles": true,
```
Comment (fenilkanjani) [inline tsconfig.json]: should not this be false?

File: tslint.json [hunk 7]
```diff
@@ -1,40 +1,26 @@
...
-    ],
-    "ordered-imports": true
+    "variable-name": [true, "allow-pascal-case"],
+    "typedef": [true, "call-signature"],
+    "file-name-casing": [true, "camel-case"],
+    "no-implicit-dependencies": [false, "dev"],
+    "only-arrow-functions": false,
+    "no-magic-numbers": false,
+    "no-parameter-properties": false,
+    "no-unused-expression": true,
```
Comment (jLouzado) [inline tslint.json]: not working for some reason, there are still unused imports all over the place and `tsc` doesn't complain. And `no-unused-variable` is deprecated, and is throwing weird errors.

File: test/smitten/smitten.ts (hunk 5 above)
Comment (jLouzado) [reply test/smitten/smitten.ts]: lint threw an error for using deprecated method `deepEqual` and suggested the change to `deepStrictEqual`

File: tslint.json (hunk 4 above)
Comment (jLouzado) [reply tslint.json]: - have activated `no-null-keyword`, and disabled it for test files
- have activated `no-magic-numbers` and disabled it for test files
- `no-parameter-properties` is set to `false` to allow constructor-param-initialization

**Pull request closed without merging**